src/directives/repeat.ts

```diff
@@ -104,6 +104,34 @@ export const repeat =
                 if (!(containerPart instanceof NodePart)) {
                   throw new Error('repeat can only be used in text bindings');
                 }
+
+                // Performing client-side hydration.
+                if (containerPart.options.prerenderedParts !== undefined) {
+                  const nodeParts: NodePart[] = [];
+                  for (const partInfo of containerPart.options.prerenderedParts!) {
+                    // Remove prerenderedParts from options, as they only apply to
+                    // the container part.
+                    const {prerenderedParts, ...options} = containerPart.options;
+                    const nodePart = new NodePart(options);
+                    nodePart.startNode = partInfo.startNode;
+                    nodePart.endNode = partInfo.endNode;
+                    nodeParts.push(nodePart);
+                  }
+                  partListCache.set(containerPart, nodeParts);
+
+                  // Hydrate the template keys only if the data is unchanged since
+                  // pre-rendering.
+                  if (!containerPart.options.dataChanged) {
+                    const keys = [];
+                    let index = 0;
+                    for (const item of items) {
+                      keys.push(keyFn ? keyFn(item, index) : index);
+                      index++;
```

Comment:
Maybe combine these loops? We skip items.length checks of `!containerPart.options.dataChanged` by keeping them separate.